File: frontend/src/api/rmm.ts
import apiClient from "./client";

export type RmmType = "tactical_rmm" | "ninja_rmm" | "atera" | "connectwise_automate";

export interface RmmIntegration {
  id: string;
  name: string;
  rmm_type: RmmType;
  base_url: string;
  verify_ssl: boolean;
  is_active: boolean;
  last_sync_at: string | null;
  last_sync_status: string | null;
  last_sync_message: string | null;
  agent_count: number;
  site_filter: string | null;
  created_at: string;
}

export interface RmmAgent {
  id: string;
  integration_id: string;
  external_id: string;
  hostname: string;
  os_name: string | null;
  ip_address: string | null;
  status: string;
  last_seen: string | null;
  patches_pending: number | null;
  alerts_count: number;
  synced_at: string;
  raw_data?: Record<string, unknown> | null;
}

export interface RmmScriptRun {
  id: string;
  integration_id: string;
  agent_external_id: string;
  agent_hostname: string;
  run_type: "script" | "command";
  shell: string;
  body: string;
  output: string | null;
  exit_code: number | null;
  status: "pending" | "running" | "success" | "error";
  started_at: string;
  finished_at: string | null;
}

export const rmmApi = {
  list: () => apiClient.get<RmmIntegration[]>("/rmm").then((r) => r.data),

  create: (data: {
    name: string;
    rmm_type: RmmType;
    base_url: string;
    credentials: Record<string, string>;
    verify_ssl?: boolean;
    site_filter?: string | null;
  }) => apiClient.post<RmmIntegration>("/rmm", data).then((r) => r.data),

  update: (id: string, data: Partial<RmmIntegration & { credentials: Record<string, string> }>) =>
    apiClient.patch<RmmIntegration>(`/rmm/${id}`, data).then((r) => r.data),

  delete: (id: string) => apiClient.delete(`/rmm/${id}`),

  test: (id: string) =>
    apiClient.post<{ ok: boolean; message: string }>(`/rmm/${id}/test`).then((r) => r.data),

  sync: (id: string) =>
    apiClient.post<{ synced: number; message: string }>(`/rmm/${id}/sync`).then((r) => r.data),

  agents: (id: string, status?: string) =>
    apiClient.get<RmmAgent[]>(`/rmm/${id}/agents`, { params: status ? { status } : undefined }).then((r) => r.data),

  run: (integrationId: string, agentExternalId: string, data: {
    run_type: "script" | "command";
    shell: string;
    body: string;
    timeout: number;
  }) =>
    apiClient
      .post<RmmScriptRun>(`/rmm/${integrationId}/agents/${agentExternalId}/run`, data)
      .then((r) => r.data),

  scriptRuns: (integrationId: string, agentId?: string) =>
    apiClient
      .get<RmmScriptRun[]>(`/rmm/${integrationId}/script-runs`, {
        params: agentId ? { agent_id: agentId } : undefined,
      })
      .then((r) => r.data),
};

export type TemplateCategory = "monitoring" | "security" | "maintenance" | "network" | "general" | "incident_response" | "identity" | "compliance" | "forensics";
export type TemplateShell = "powershell" | "cmd" | "python" | "bash";

export interface RmmScriptTemplate {
  id: string;
  tenant_id: string | null;
  name: string;
  description: string | null;
  category: TemplateCategory;
  shell: TemplateShell;
  run_type: "command" | "script";
  body: string;
  is_builtin: boolean;
  created_at: string;
}

export const rmmTemplatesApi = {
  list: (category?: string) =>
    apiClient
      .get<RmmScriptTemplate[]>("/rmm/templates", { params: category ? { category } : undefined })
      .then((r) => r.data),

  seed: () =>
    apiClient.post<{ added: number; message: string }>("/rmm/templates/seed").then((r) => r.data),

  create: (data: {
    name: string;
    body: string;
    shell: string;
    run_type: string;
    category: string;
    description?: string;
  }) => apiClient.post<RmmScriptTemplate>("/rmm/templates", data).then((r) => r.data),

  update: (id: string, data: Partial<Omit<RmmScriptTemplate, "id" | "tenant_id" | "is_builtin" | "created_at">>) =>
    apiClient.patch<RmmScriptTemplate>(`/rmm/templates/${id}`, data).then((r) => r.data),

  delete: (id: string) => apiClient.delete(`/rmm/templates/${id}`),
};

export const ssoMappingsApi = {
  list: () => apiClient.get("/sso-mappings").then((r) => r.data),
  create: (data: { external_group: string; platform_role: string }) =>
    apiClient.post("/sso-mappings", data).then((r) => r.data),
  delete: (id: string) => apiClient.delete(`/sso-mappings/${id}`),
  roles: () => apiClient.get<{ roles: string[] }>("/sso-mappings/roles").then((r) => r.data),
};

export const fileSharesApi = {
  list: () => apiClient.get("/file-shares").then((r) => r.data),
  create: (data: object) => apiClient.post("/file-shares", data).then((r) => r.data),
  delete: (id: string) => apiClient.delete(`/file-shares/${id}`),
  getScript: (id: string) => apiClient.get<{ script: string; language: string }>(`/file-shares/${id}/script`).then((r) => r.data),
  submitScan: (id: string, data: { shares: object[]; acls: object[] }) =>
    apiClient.post(`/file-shares/${id}/scan-result`, data).then((r) => r.data),
  shares: (id: string) => apiClient.get(`/file-shares/${id}/shares`).then((r) => r.data),
  acls: (configId: string, shareId: string) =>
    apiClient.get(`/file-shares/${configId}/shares/${shareId}/acls`).then((r) => r.data),
};

export const syslogApi = {
  list: () => apiClient.get("/siem-syslog").then((r) => r.data),
  create: (data: object) => apiClient.post("/siem-syslog", data).then((r) => r.data),
  update: (id: string, data: object) => apiClient.patch(`/siem-syslog/${id}`, data).then((r) => r.data),
  delete: (id: string) => apiClient.delete(`/siem-syslog/${id}`),
  test: (id: string) =>
    apiClient.post<{ ok: boolean; message: string }>(`/siem-syslog/${id}/test`).then((r) => r.data),
};

File: backend/app/models/rmm.py
"""ORM models — F23.ext RMM Integrations."""
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID, uuid4

from sqlalchemy import Boolean, ForeignKey, Integer, String, Text, TIMESTAMP
from sqlalchemy.dialects.postgresql import JSONB, UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.database import Base


class RmmIntegration(Base):
    __tablename__ = "rmm_integrations"

    id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    rmm_type: Mapped[str] = mapped_column(String(30), nullable=False)
    base_url: Mapped[str] = mapped_column(Text, nullable=False)
    config_encrypted: Mapped[str | None] = mapped_column(Text, nullable=True)
    verify_ssl: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    last_sync_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)
    last_sync_status: Mapped[str | None] = mapped_column(String(20), nullable=True)
    last_sync_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    agent_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    site_filter: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    agents: Mapped[list["RmmAgent"]] = relationship("RmmAgent", back_populates="integration", cascade="all, delete-orphan")
    script_runs: Mapped[list["RmmScriptRun"]] = relationship("RmmScriptRun", back_populates="integration", cascade="all, delete-orphan")


class RmmAgent(Base):
    __tablename__ = "rmm_agents"

    id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True, default=uuid4)
    integration_id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), ForeignKey("rmm_integrations.id", ondelete="CASCADE"), nullable=False, index=True)
    tenant_id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False, index=True)
    external_id: Mapped[str] = mapped_column(String(200), nullable=False)
    hostname: Mapped[str] = mapped_column(String(200), nullable=False)
    os_name: Mapped[str | None] = mapped_column(Text, nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(50), nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="unknown")
    last_seen: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)
    patches_pending: Mapped[int | None] = mapped_column(Integer, nullable=True)
    alerts_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    raw_data: Mapped[Any | None] = mapped_column(JSONB, nullable=True)
    synced_at: Mapped[datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=func.now(), nullable=False)

    integration: Mapped["RmmIntegration"] = relationship("RmmIntegration", back_populates="agents")


class RmmScriptRun(Base):
    __tablename__ = "rmm_script_runs"

    id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True, default=uuid4)
    integration_id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), ForeignKey("rmm_integrations.id", ondelete="CASCADE"), nullable=False, index=True)
    tenant_id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False, index=True)
    agent_external_id: Mapped[str] = mapped_column(String(200), nullable=False)
    agent_hostname: Mapped[str] = mapped_column(String(200), nullable=False)
    run_type: Mapped[str] = mapped_column(String(20), nullable=False)
    shell: Mapped[str] = mapped_column(String(20), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    output: Mapped[str | None] = mapped_column(Text, nullable=True)
    exit_code: Mapped[int | None] = mapped_column(Integer, nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")
    executed_by: Mapped[UUID | None] = mapped_column(PG_UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    started_at: Mapped[datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=func.now(), nullable=False)
    finished_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)

    integration: Mapped["RmmIntegration"] = relationship("RmmIntegration", back_populates="script_runs")


class RmmScriptTemplate(Base):
    __tablename__ = "rmm_script_templates"

    id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID | None] = mapped_column(PG_UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=True, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    category: Mapped[str] = mapped_column(String(50), nullable=False, default="general")
    shell: Mapped[str] = mapped_column(String(20), nullable=False, default="powershell")
    run_type: Mapped[str] = mapped_column(String(10), nullable=False, default="command")
    body: Mapped[str] = mapped_column(Text, nullable=False)
    is_builtin: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
